### DISCUSSION TITLE
`$type` in inferred semantic model

### DISCUSSION THREAD

**Discussion by coolya** [eclipse-langium/langium#769] [Ideas]
The inferred model from a grammar doesn't further constrain the type of the `$type` property. It is still `string` while the `$container` is constrained to the possible parents derived from the grammar. 

Is there a reason why: 

```
export interface BooleanLiteral extends AstNode {
    readonly $container: ...;
    readonly $type: 'BooleanLiteral';
    value: boolean
}

export const BooleanLiteral = 'BooleanLiteral';
```

Wouldn't work? 

Especially in testing where you don't always want to construct an AST from the concrete syntax,  it would make direct construction of the AST much simpler when the type would have a constraint.

**Comment by msujew** [comment]
We've played around with this idea as well, and in theory it's possible. It's a bit complicated once we add some level of inheritance, as then the `$type` property needs to reflect every possible option of subtypes, but we already do this for the `$container` types anyway.

It might be worth looking into this in the context of the planned type collection refactoring. @pluralia @montymxb your thoughts?

**Reply by pluralia** [reply]
I'll add it in the current round of type collection refactoring.
